### PR TITLE
Smarter Choice AI for Status Moves

### PR DESCRIPTION
**Pull request by Pawkkie**
## Description
Earlier this week @wiz1989 let me know about some poor AI behaviour they'd seen in Pit of Trials. Randomized items lead to a mon with a Choice item using Yawn instead of a damaging move, and never switching out. A bit of investigating made it clear that the AI doesn't have anything to handle a mon with a choice item and a status move, and it would spam even setup moves like Swords Dance. This PR aims to make choice items and status moves work together in a reasonable manner.

I've started writing tests first in my PRs to help define the scope of what exactly I want to do. It's also really helpful to describe exactly what behaviour I've added. This PR was intended to have the AI pass the following tests:
- Choiced Pokémon switch out after using a status move once
- Choiced Pokémon won't use stat boosting moves *(really moves with MOVE_TARGET_USER)
- Choiced Pokémon won't use a status move if they are the only party member
- Choiced Pokémon won't use a status move if they don't have a good switchin
- Choiced Pokémon won't use a status move if they are trapped

The reason for this approach rather than just blanket disallowing the AI from using status moves with a choice item altogether is to let the AI still be able to use sets like the classic scarfed Rotom-W with Will-O-Wisp or Landorus-T with Stealth Rock.

I should also note that the section of code that's in the switch AI (for the first bullet point) does _not_ require any smart switching flags to work. I figure this is reasonable default behaviour for a non-vanilla item without needing to gate it behind a flag. Otherwise, the rest of the changes are tied to the AI_FLAG_CHECK_BAD_MOVE flag, as I thought that was the most representative.

The * on the 2nd bullet point is any exceptions for moves with MOVE_TARGET_USER. I'm happy to expand this list if there are other suggestions. Currently, exceptions to this blanket ban are Destiny Bond, Wish, Healing Wish, Lunar Dance, and Aurora Veil in the appropriate weather.

## **People who collaborated with me in this PR**
Wiz, for discovering the behaviour

## **Discord contact info**
@Pawkkie

### CODE REVIEW

File: src/battle_ai_main.c [hunk 1]
```diff
@@ -2654,6 +2656,23 @@ static s32 AI_CheckBadMove(u32 battlerAtk, u32 battlerDef, u32 move, s32 score)
             return 0;   // cannot even select
     } // move effect checks
 
+    // Choice items
+    if (gBattleMons[battlerAtk].ability != ABILITY_KLUTZ
```
Comment (DizzyEggg) [inline src/battle_ai_main.c:2660]: Small nitpick, but I would change the order here, and first check holdEffect, then ability

File: src/battle_ai_main.c [hunk 2]
```diff
@@ -2654,6 +2656,23 @@ static s32 AI_CheckBadMove(u32 battlerAtk, u32 battlerDef, u32 move, s32 score)
...
+    && (holdEffect == HOLD_EFFECT_CHOICE_SPECS || holdEffect == HOLD_EFFECT_CHOICE_BAND || holdEffect == HOLD_EFFECT_CHOICE_SCARF))
+    {
+        // Don't use user-target moves ie. Swords Dance, with exceptions
+        if ((moveTarget & MOVE_TARGET_USER)
+        && move != MOVE_DESTINY_BOND && move != MOVE_WISH && move != MOVE_HEALING_WISH && move != MOVE_LUNAR_DANCE
+        && !(move == MOVE_AURORA_VEIL && ((AI_GetWeather(aiData) & B_WEATHER_SNOW) || (AI_GetWeather(aiData) & B_WEATHER_HAIL))))
+            ADJUST_SCORE(-30);
+        // Don't use a status move if the mon is the last one in the party, has no good switchin, or is trapped
+        else if (moveCategory == DAMAGE_CATEGORY_STATUS
```
Comment (DizzyEggg) [inline src/battle_ai_main.c:2669]: I would just directly call GetBattleMoveCategory

File: src/battle_ai_main.c [hunk 3]
```diff
@@ -2654,6 +2656,23 @@ static s32 AI_CheckBadMove(u32 battlerAtk, u32 battlerDef, u32 move, s32 score)
             return 0;   // cannot even select
     } // move effect checks
 
+    // Choice items
+    if (gBattleMons[battlerAtk].ability != ABILITY_KLUTZ
+    && (holdEffect == HOLD_EFFECT_CHOICE_SPECS || holdEffect == HOLD_EFFECT_CHOICE_BAND || holdEffect == HOLD_EFFECT_CHOICE_SCARF))
+    {
+        // Don't use user-target moves ie. Swords Dance, with exceptions
+        if ((moveTarget & MOVE_TARGET_USER)
+        && move != MOVE_DESTINY_BOND && move != MOVE_WISH && move != MOVE_HEALING_WISH && move != MOVE_LUNAR_DANCE
```
Comment (DizzyEggg) [inline src/battle_ai_main.c:2665]: I would check for battle move's effect instead of move id

File: src/battle_ai_main.c [hunk 4]
```diff
@@ -2654,6 +2656,23 @@ static s32 AI_CheckBadMove(u32 battlerAtk, u32 battlerDef, u32 move, s32 score)
...
+        && !(move == MOVE_AURORA_VEIL && ((AI_GetWeather(aiData) & B_WEATHER_SNOW) || (AI_GetWeather(aiData) & B_WEATHER_HAIL))))
```
Comment (DizzyEggg) [inline src/battle_ai_main.c:2666]:
```C
&& !(move == MOVE_AURORA_VEIL && (AI_GetWeather(aiData) & (B_WEATHER_SNOW | B_WEATHER_HAIL)) || 
```

File: src/battle_ai_switch_items.c [hunk 5]
```diff
@@ -919,6 +919,26 @@ static bool32 ShouldSwitchIfEncored(u32 battler, bool32 emitResult)
...
+        && gBattleMons[battler].ability != ABILITY_KLUTZ)
+    {
+        if (gMovesInfo[gLastUsedMove].category == DAMAGE_CATEGORY_STATUS)
+        {
+            gBattleStruct->AI_monToSwitchIntoId[battler] = PARTY_SIZE;
+            if (emitResult)
+                BtlController_EmitTwoReturnValues(battler, 1, B_ACTION_SWITCH, 0);
+            return TRUE;
+        }
+
```
Comment (DizzyEggg) [inline src/battle_ai_switch_items.c:936]: empty line

File: src/battle_ai_main.c [hunk 6]
```diff
@@ -2654,6 +2656,23 @@ static s32 AI_CheckBadMove(u32 battlerAtk, u32 battlerDef, u32 move, s32 score)
             return 0;   // cannot even select
     } // move effect checks
 
+    // Choice items
+    if (gBattleMons[battlerAtk].ability != ABILITY_KLUTZ
+    && (holdEffect == HOLD_EFFECT_CHOICE_SPECS || holdEffect == HOLD_EFFECT_CHOICE_BAND || holdEffect == HOLD_EFFECT_CHOICE_SCARF))
```
Comment (DizzyEggg) [inline src/battle_ai_main.c:2661]: Use `HOLD_EFFECT_CHOICE` macro

File: src/battle_ai_switch_items.c [hunk 7]
```diff
@@ -919,6 +919,26 @@ static bool32 ShouldSwitchIfEncored(u32 battler, bool32 emitResult)
     return FALSE;
 }
 
+static bool32 ShouldSwitchIfBadChoiceLock(u32 battler, bool32 emitResult)
+{
+    u32 holdEffect = ItemId_GetHoldEffect(gBattleMons[battler].item);
+
+    if ((holdEffect == HOLD_EFFECT_CHOICE_SPECS || holdEffect == HOLD_EFFECT_CHOICE_BAND || holdEffect == HOLD_EFFECT_CHOICE_SCARF)
```
Comment (DizzyEggg) [inline src/battle_ai_switch_items.c:926]: Use `HOLD_EFFECT_CHOICE` macro

File: src/battle_ai_switch_items.c [hunk 8]
```diff
@@ -919,6 +919,26 @@ static bool32 ShouldSwitchIfEncored(u32 battler, bool32 emitResult)
     return FALSE;
 }
 
+static bool32 ShouldSwitchIfBadChoiceLock(u32 battler, bool32 emitResult)
+{
+    u32 holdEffect = ItemId_GetHoldEffect(gBattleMons[battler].item);
```
Comment (DizzyEggg) [inline src/battle_ai_switch_items.c:924]: Use `GetBattlerHoldEffect`

File: test/battle/hold_effect/choice.c [hunk 9]
```diff
@@ -0,0 +1,181 @@
+#include "global.h"
```
Comment (DizzyEggg) [inline test/battle/hold_effect/choice.c:1]: This folder is usually for battle tests, I would put it with other ai battle tests for now, and then we could make a seperate PR to create a new AI folder for ai-related tests.

**Review by DizzyEggg — CHANGES_REQUESTED** [review]
Awesome work!

File: test/battle/hold_effect/choice.c (hunk 9 above)
Comment (Pawkkie) [reply test/battle/hold_effect/choice.c:1]: Super on board for that idea! :D

File: src/battle_ai_main.c [hunk 10]
```diff
@@ -806,6 +806,7 @@ static s32 AI_CheckBadMove(u32 battlerAtk, u32 battlerDef, u32 move, s32 score)
     u32 i;
     u32 weather;
     u32 predictedMove = aiData->predictedMoves[battlerDef];
+    u32 holdEffect = ItemId_GetHoldEffect(gBattleMons[battlerAtk].item);
```
Comment (DizzyEggg) [inline src/battle_ai_main.c:809]: ACTUALLY, you can get rid of this variable I think

File: src/battle_ai_main.c [hunk 11]
```diff
@@ -2654,6 +2655,22 @@ static s32 AI_CheckBadMove(u32 battlerAtk, u32 battlerDef, u32 move, s32 score)
             return 0;   // cannot even select
     } // move effect checks
 
+    // Choice items
+    if (HOLD_EFFECT_CHOICE(holdEffect) && gBattleMons[battlerAtk].ability != ABILITY_KLUTZ)
```
Comment (DizzyEggg) [inline src/battle_ai_main.c:2659]: And instead do
`HOLD_EFFECT_CHOICE(aiData->holdEffects[battlerAtk]) `

**Review by DizzyEggg — CHANGES_REQUESTED** [review]
That's the last thing and I'll merge, I promise 😂

**Comment by Pawkkie** [comment]
> That's the last thing and I'll merge, I promise 😂

LOL

No worries whatsoever, thanks as always for the feedback! :D

**Review by DizzyEggg — APPROVED** [review]

**Pull request merged**